Step4. 자동차경주(우승자) 리뷰 요청드립니다.

AyoungLim1103 opened the pull request:
- input a name of car.(the length should be under 5)
- compare each position of cars to calculate the max position.
- when print result track, print the name of car.

Commit:
feat(racingCar) : Make a new name for each car for race and print the name of winners.

- input a name of car.(the length should be under 5)
- compare each position of cars to calculate the max position.
- when print result track, print the name of car.

AyoungLim1103 committed:
Update README.md

gw-study commented on README.md:
```diff
@@ -12,3 +12,8 @@
 3. 전진 여부는 random 함수를 이용하여 random value 가 4이상 9이하일 경우만 전진하도록 설정한다.
 4. 결과는 '-' (대시) 를 통해 보여준다.
+
+## 2차 기능 구현
+1. 자동차경주를 시작할 자동차에 각 이름을 붙여준다. (단, 이름은 5자 이내)
+2. 자동차 경주 결과 출력시 자동차의 이름도 함께 보여준다.
+3. 자동차 경주가 끝난 후 우승자의 이름을 출력한다.
```
기능목록 정리👍

gw-study commented on src/main/java/racingcar/ResultView.java:
```diff
@@ -10,9 +10,13 @@ public void startResult(){
+            System.out.print(car.getName() + " :");
             System.out.println(car.showTrack(TRACKSHAPE));
         }
+        System.out.println("");
```
```suggestion
        System.out.println();
```
이렇게 변경가능합니다😀

gw-study commented on src/main/java/racingcar/Name.java:
```diff
@@ -0,0 +1,16 @@
+package racingcar;
+
+public class Name {
```
자동차 이름을 클래스로 분리하셨군요👍
객체지향 생활 제조 원칙 `규칙 3: 모든 원시값과 문자열을 포장한다.`을 잘 활용해주셨어요💯

gw-study commented on src/main/java/racingcar/Name.java:
```diff
@@ -0,0 +1,16 @@
+        if(name.length() > MAX_NAME_LENGTH)
+            throw new IllegalArgumentException("자동차 이름은 5자 이하여야 합니다.");
+
+        this.name = name.trim();
```
공백을 먼저 자른 다음에 검증 체크를 하는건 어떨까요?🤔
검증을 통과한 값에 대한 수정은 가능하면 안 하는 쪽이 더 좋을거 같습니다 :)

gw-study commented on src/main/java/racingcar/Car.java:
```diff
@@ -11,6 +17,19 @@ public int moveForward(int value){
+    }
+
+    public int maxPosition(int lastPosition){
+        if(currentPosition > lastPosition)
+            return currentPosition;
+        return lastPosition;
```
조금 더 간단하게 변경해 보면 아래처럼 가능하겠네요😁
```suggestion
        return currentPosition > lastPosition ? currentPosition : lastPosition;
```
더 간단히 하면 요렇게요😁
```suggestion
        return Math.max(currentPosition, lastPosition);
```

gw-study commented on src/main/java/racingcar/Car.java:
```diff
@@ -11,6 +17,19 @@ public int moveForward(int value){
+        return lastPosition;
+    }
+
     public String showTrack(char trackShape) {
         StringBuffer track = new StringBuffer();
```
`showTrack`함수는 이름에서도 알 수 있듯이 출력을 담당하는 부분인거 같아요
View쪽으로 이동 부탁드립니다🙇‍♂️

gw-study commented on src/main/java/racingcar/Racing.java:
```diff
@@ -4,24 +4,44 @@
-        for(int i = 0; i < CarCount; i++){
-            cars.add(new Car());
+    public void readyCar(String[] cars){
+        for(int i = 0; i < cars.length; i++){
+            readyCars.add(new Car(cars[i], 0));
```
```suggestion
        for (String car : cars) {
            readyCars.add(new Car(car, 0));
```
이렇게 변경하면 코드 읽기가 더 쉽겠네요😁

gw-study commented on src/main/java/racingcar/Racing.java:
```diff
@@ -4,24 +4,44 @@
+    }
+
+    public int getMaxPosition(){
+        int position = readyCars.get(0).getPosition();
```
0은 매직넘버 입니다. 매직넘버 사용을 지양해보도록 해보세요🙏

AyoungLim1103 replied:
int maxpostion = 0; 으로 변수 선언하여 크기 비교하였는데 이렇게 변수로 선언하는것도 매직넘버가 되는지 문의드립니다....!

gw-study replied:
사실 말씀 주신 케이스도 매직넘버에 해당되긴 합니다.
매직 넘버라는게 내가 아닌(혹은 시간이 흐른뒤에) 다른 사람이 봤을때 그 의미를 해석해야
하는 경우가 대표적이라고 할 수 있을거 같아요.
예를 들어주신 경우는 일반적으로 그냥 그대로 사용하기 때문에 별 문제는 없긴 하지만
이왕이면 `int maxpostion = INIT_POSITION_NUM`;  과 같이 사용하면
코드를 읽는데 좀 더 수월할 거 같아요😊

지금과 같은 경우는 사실 `0`이 문제라기 보단 여러번 `.`을 사용하고 있는 부분이
코드의 의미를 파악하는데 좀 방해가 되는 부분입니다😂

gw-study commented on src/main/java/racingcar/Racing.java:
```diff
@@ -4,24 +4,44 @@
+        return winners;
+    }
+    public void addWinners(Car car, Winners winners){
+        if(car.getPosition() == getMaxPosition())
+            winners.addWinner(car);
```
자바 코드 컨벤션을 지켜주도록 해주세요🙏
```suggestion
        if (car.getPosition() == getMaxPosition()) { 
            winners.addWinner(car);   
        }
```

gw-study commented on src/main/java/racingcar/Racing.java:
```diff
@@ -4,24 +4,44 @@
+
+    public Winners getWinners(){
+        Winners winners = new Winners();
+        for(Car car : readyCars){
```
포맷팅 적용 부탁드립니다🙏

AyoungLim1103 replied:
차장님,  피드백주신 for 반복문에서 어느부분을 포맷팅 적용할 수 있는지 찾지 못하여 문의드립니다 ㅠㅠ....

gw-study replied:
```java
    public Winners getWinners() {
        Winners winners = new Winners();
        for (Car car : readyCars) {
               ......
        }
    }
```
이렇게 사용하시는게 [코드 컨벤션](https://naver.github.io/hackday-conventions-java/)에 부합됩니다.
인텔리제이를 사용한다면 `Save Actions`로 자동화 해 볼 수 있어요😁

gw-study commented on src/main/java/racingcar/Racing.java:
```diff
@@ -4,24 +4,44 @@
 
     public List<Car> moveCars(){
-        for(Car car : cars){
+        for(Car car : readyCars){
```
포맷팅 적용 부탁드립니다🙏

AyoungLim1103 replied:
차장님,  위 케이스와 마찬가지로 해당 부분에서도 어느부분을 포맷팅 적용할 수 있는지 찾지 못하여 문의드립니다 ㅠㅠ.......

gw-study replied:
```java
public List<Car> moveCars() {
        for (Car car : readyCars) {
            .....
        }
    }
```
이것도 요렇게 되겠죠?😁

gw-study commented on src/main/java/racingcar/Winners.java:
```diff
@@ -0,0 +1,26 @@
+    public Winners(){
+        this(new ArrayList<>());
+    }
+    public Winners(List<Car> cars){
+        this.winners = cars;
+    }
```
생성자 초기화를 잘 사용하셨네요👍

gw-study commented on src/main/java/racingcar/Winners.java:
```diff
@@ -0,0 +1,26 @@
+    public void addWinner(Car car){
+        this.winners.add(car);
+    }
+
+    public String winnerNames(){
```
우승자를 조회하고 우승자들의 출력을 위해 `,`을 붙여주는 역할은
View에서 해 주어야 할 역할 같습니다😂
`Winners`는 우승자들만 가지고 있고 그와 관련된 행위만 가지고 있는게 좋을거 같아요 :)

gw-study commented on src/test/java/racingcar/RacingCarTest.java:
```diff
@@ -9,12 +9,12 @@
 
 public class RacingCarTest {
```
아래와 같은 테스트를 추가해 볼 수 있을까요?
```
자동차 = A,B,C 3대
시도횟수 = 2회 라고 하면,
--1회
a.move(1);
b.move(2);
c.move(8);

--2회
a.move(1);
b.move(5);
c.move(6);

우승자는 A 입니다
```
이런 느낌으로 자동차별 시도횟수를 제어하여 우승자 예측이 가능한 테스트를 해볼 수 있으면 좋을거 같아요😄

AyoungLim1103 replied:
말씀주신대로 테스트케이스를 작성하였는데 적절한 케이스구문인지 확신이 안서네요 ㅠㅠ;;;

gw-study commented on src/main/java/racingcar/Car.java:
```diff
@@ -3,6 +3,12 @@
 public class Car {
     private int currentPosition = 0;
     private static final int FOUR = 4;
```
상수 선언이 처음에 오고 그 다음 멤버 변수 선언이 오는것이 좋다고
코드 컨벤션에 있는거 같아요😂

gw-study reviewed (changes requested):
안녕하세요😄
주말에도 열심히 과정 진행해주셨네요👍

자동차 경주 우승자가 아주 잘 출력되도록 잘해주셨어요👏
그치만 잘 출력이 되는게 끝이아니기에...
코멘트를 몇개 달아놨습니다😂

이번 단계의 핵심 내용은
 - 각 클랙스 및 메소드의 역할 분리
 - 테스트가 가능한 코드와 테스트가 힘든 코드의 분리
 - 단위 테스트 방법 확인
 - 기타 커밋, 코드 컨벤션 지키기
이 정도 되지 않나 생각됩니다.

핵심 내용이 잘 학습되고 있는지에 대해 저도 코멘트를 달아 드리고 있는데요,
제가 빠트린 부분이 있을수도 있으니 아영님도 요런 내용을 고민할 수 있는 시간을 가져보는 것도 좋을거 같습니다 :)
고민하다가 막히는 부분이 있으면 언제든지 연락주셔도 괜찮구요😄

그럼 다음 리뷰요청 기다리겠습니다
남은 주말 잘 보내세요~😎

Commit:
fix(racingCar) : fix a test case, follow some code convention rule.

- fix test case for getting names of winners.
- follow some code convention ruels.
(ex, varialbe order..)
- make a function more simple.

Commit:
Merge remote-tracking branch 'origin/step4' into step4

Commit:
fix(racingCar) : fix a test case, follow some code convention rule.

- fix test case for getting names of winners.
- follow some code convention ruels.
(ex, varialbe order..)
- make a function more simple.

AyoungLim1103 commented:
차장님, 피드백 주신 부분 반영하여 2차 리뷰 요청드립니다~! 
단, (1) 포맷팅 (2) 매직넘버 사용 지양 (3) 테스트 케이스 적용  부분은 제가 헷갈린 부분이 있어서 댓글로 문의남겼습니다!!
감사합니다~!ㅎㅎ

gw-study commented on src/main/java/racingcar/Car.java:
```diff
@@ -1,8 +1,14 @@
 
 public class Car {
-    private int currentPosition = 0;
     private static final int FOUR = 4;
```
상수도 의미가 있는 이름을 가지면 좋을거 같아요
예를 들면 `MOVABLE_NUMBER`정도면 좋을까요?😄

gw-study commented on src/test/java/racingcar/RacingCarTest.java:
```diff
@@ -33,7 +34,16 @@ public void moveForward(){
     @DisplayName("자동차 위치 표시")
     public void showRacingCarResult() {
```
테스트가 실패하네요😱
![image](https://user-images.githubusercontent.com/84048349/123953878-b22c0e00-d9e2-11eb-8c29-185968a31ba0.png)

gw-study commented on src/main/java/racingcar/ResultView.java:
```diff
@@ -9,10 +9,22 @@ public void startResult(){
         }
+        System.out.println();
+    }
+    public static void printWinners(Winners winners){
```
`static`키워드는 필요 없을거 같은 생각인데 어떨까요?

gw-study commented on src/main/java/racingcar/Racing.java:
```diff
@@ -4,24 +4,45 @@
 import java.util.List;
 
 public class Racing {
```
생성자를 이용해서 멤버변수를 초기화하면
멤버변수의 값도 보장할 수 있고 클래스의 역할도 명확해지는 장점이 있습니다😄
초기화 부분만 아래처럼 변경해 봤습니다. 어떤게 괜찮을까요?🤔
```java
public class Racing {
    private final List<Car> readyCars;

    public Racing(String[] cars) {
        this.readyCars = readyCar(cars);
    }

    private List<Car> readyCar(String[] cars) {
        List<Car> racingCars = new ArrayList<>();

        for (String car : cars) {
            racingCars.add(new Car(car, 0));
        }

        return racingCars;
    }
}
```

gw-study commented on src/main/java/racingcar/Name.java:
```diff
@@ -0,0 +1,15 @@
+    private final String name;
+
+    public Name(String name){
+        this.name = name.trim();
+        if(name.length() > MAX_NAME_LENGTH)
```
피드백 적용 감사합니다👍

gw-study commented on src/main/java/racingcar/Car.java:
```diff
@@ -11,12 +17,14 @@ public int moveForward(int value){
-        }
-        return track.toString();
+    public int maxPosition(int lastPosition){
+        return Math.max(currentPosition, lastPosition);
```
간결하게 표현해주셨네요👍

gw-study commented on src/main/java/racingcar/ResultView.java:
```diff
@@ -9,10 +9,22 @@ public void startResult(){
 
     private static final char TRACKSHAPE = '-';
```
앗!제가 저번 리뷰에서 놓쳤나 봅니다😂
상수 선언은 클래스의 맨첫부분에서 해주세요🙏

gw-study commented on src/main/java/racingcar/Racing.java:
```diff
@@ -4,24 +4,45 @@
+    }
+
+    public int getMaxPosition(){
+        int maxPosition = 0;
```
이 정도 매직넘버는 괜찮습니다. 누구나가 다 이해할 수 있으니까요😄
참고로 루프문 안에 있는 `i = 0` 과 같은 것도 매직넘버로 보긴 힘들거 같아요

gw-study commented on src/test/java/racingcar/RacingCarTest.java:
```diff
@@ -33,7 +34,16 @@ public void moveForward(){
+
+    @Test
+    @DisplayName("우승자 선정 테스트")
+    public void getWinnersName(){
```
테스트 작성 잘해주셨어요👏👏👏

다만 바램이 있다면 점`.`을 좀 줄일 수 있을까요?
`racing.getCars().get(0).moveForward(1);` 이건 점이 3개나 들어있어요😭

제 욕심엔
```java
Car car1 = new Car("crong", 3);
Car car2 = new Car("pobi", 2);

car1.moveForward(4); //현재위치 4
car2.moveForward(5); //현재위치 2
```
이렇게 되어봤으면 좋겠는데요 꼭 수정하시지는 않아도 괜찮습니다

gw-study reviewed (changes requested):
빠르게 피드백 적용해주셨는데
제가 너무 늦게 리뷰를 해드렸네요. 죄송합니다😭

댓글로 궁금한 내용에 대해 제 나름대로 코멘트를 드렸는데요 도움이 되셨을지 모르겠네요
사실, 저도 항상 고민하는 내용들이라 제 의견이 맞다고는 말씀 못 드리겠어요😁

코드 잘 작성해주셨는데요
제가 미션을 하나 드리고 싶은게 있습니다 :)
이건 꼭 수행하지 않아도 괜찮으니까 도전 정도는 한번 해봐주세요.

>현재 우승자를 구하는 메소드가 일부분 `Racing`객체에 포함되어 있습니다.
단일책임원칙(SRP)에 위반 소지가 있는 부분인데요,
아래 제시해 드린 것처럼 변경해 볼 수 있을까요?
>
>우승자를 구하는 Winners클래스는
>- `cars`(경주 자동차들)을 인스턴스 변수로 가진다.
>- 인스턴스변수는 생성자 초기화로 초기화 한다.
>- `findWinnerCarNames`(우승 자동차 이름들 구하기) 메소드를 가진다
>- `getMaxPosition`은 여기서 구현하고 우승자를 구할때 사용한다

이렇게 해볼 수 있으면 좋겠습니다!도전!?😀

그외 테스트가 실패하는 부분을 포함해서
몇가지 피드백 드린 내용이 있으니 확인 부탁드릴께요🙇‍♂️

다음 리뷰 요청 기다리겠습니다
(리뷰 요청 시 DM으로 알려주시면 제가 안 까먹구 빨리 할 수 있을거 같아요😁)

Commit:
fix(racingCar) : fix a test case, follow some code convention rule.

- fix test case for getting names of winners.
- follow some code convention ruels.
(ex, varialbe order..)
- fix Racing.java function.

Commit:
fix(racingCar) : fix a test case, follow some code convention rule.

- fix test case for getting names of winners.
- follow some code convention ruels.
(ex, varialbe order..)
- fix Racing.java function.

AyoungLim1103 commented:
차장님 안녕하세요 ㅎㅎ 상세하게 피드백 주셔서 정말 감사합니다~~!
코멘트 달아주신 부분 모두 반영하여 다시 push 했습니다! 리뷰 부탁드리겠습니다 :)
('우승자 선정 테스트' 케이스를 수정하였는데 테스트 케이스 내용이 좀 길어진거같아서 적절하게 작성한건지 모르겠습니다 ㅠㅠ..)

gw-study commented on src/test/java/racingcar/RacingCarTest.java:
```diff
@@ -1,39 +1,62 @@
+
+    @Test
+    @DisplayName("우승자 선정 테스트")
+    public void getWinnersName(){
```
우와!변경 정말 잘해주셨습니다👍

이왕 하는거 제가 조금만 더 힌트를 드릴께요😁

제가 피드백 드린 내용중에 Winners의 역할 강화도 있지만
기존 Racing에서 하고 있던 **우승자 찾기** 역할을 분리하는게 큰 의미였어요 :)

아래는 제가 만들어 본 테스트 코드예요.
```java
    @Test
    @DisplayName("우승자 선정 테스트")
    public void getWinnersName_뉴버젼() {
        Car car1 = new Car("pobi", 3);
        Car car2 = new Car("crong", 2);
        Car car3 = new Car("honux", 2);

        car1.moveForward(4);
        car2.moveForward(5);
        car3.moveForward(2);

        List<Car> cars = new ArrayList<>();

        cars.add(car1);
        cars.add(car2);
        cars.add(car3);

        Winners winners = new Winners(cars);

        assertThat(winners.winnerCars()).containsExactly(car1);
    }
```
우승자를 구하기 위해서 Racing 객체는 필요가 없어졌습니다!

이 코드가 동작하기 위해서 아래처럼 변경했습니다.
- Car.java : 객체 비교를 위해 `equals` `hashCode` 메소드 오버라이드(인텔리제이가 자동으로 만들어주어요👍)
```java
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Car car = (Car) o;
        return currentPosition == car.currentPosition && Objects.equals(name, car.name);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, currentPosition);
    }
```
- Winners.java : 메서드 추가(스트림을 사용하여 간단하게 표현하였습니다.)
```java
    public List<Car> winnerCars() {
        return cars.stream()
                .filter(car -> car.getPosition() == getMaxPosition(cars))
                .collect(Collectors.toList());
    }
```
- Racing.java
`getWinners` `addWinners` 메소드 삭제

이렇게 구현함으로 다른 중복된 메소드들도 정리가 가능할거란 생각이 들어요😊

gw-study commented on src/main/java/racingcar/Winners.java:
```diff
@@ -0,0 +1,41 @@
+
+public class Winners {
+    private final List<Car> cars;
+    private static final int INIT_POSITION_NUM = 0;
```
상수 위치는 어디였지요?🤔
위치 이동 부탁드리겠습니다😁

gw-study reviewed (changes requested):
모든 피드백 빠르게 적용해 주셔서 감사해요!👍

열심히 단계 진행 해 주신 덕분에 이번 단계 끝이 보이는거 같습니다!진짜로요!!😁

코드를 보다가 포맷팅 관련 부분이 눈에 띄어서 아래 적어드립니다.

- 코드를 보면 포맷팅이 일괄적이지 않은게 보입니다.
예를 들면 `for(int ...)` `if(.....)` 처럼 for문과 if문 사이 공백이 없는 곳이 종종 보여요
포맷팅 자동 변경 툴을 사용 중이시라면 다시 한번 확인해 보셔야 할 것 같습니다😃

그리고 본문에도 내용 주신 테스트 케이스를 확인해 봤는데요,
아주 잘 작성해 주셨습니다.
조금 더 변경해 보면 좋을 부분이 보여서 제가 관련된 코멘트를 남겨놨어요😂

피드백 드린 부분 한번 더 고민해보시라고 Request Change 합니다
금방 이번 단계 끝날 수 있을거 같아요!화이팅!🔥🔥🔥

Commit:
fix(racingCar) : fix a test case, follow some code convention rule, fix Winners.java

- fix test case for getting names of winners.
- follow some code convention ruels.
(ex, varialbe order..)
- fix Winners.java for getting the name of winners.
- remove duplicate classes.

AyoungLim1103 commented:
안녕하세요 차장님 : ) 피드백 빠르게 주셔서 감사합니다!!
피드백 주신대로 winner, racing 클래스 변경해보았습니다~ㅎㅎ 코드가 조금 깔끔해진거 같습니다..!ㅎㅎㅎ
리뷰 부탁드리겠습니다 :) 감사합니다~!ㅎㅎ

gw-study commented on src/main/java/racingcar/Car.java:
```diff
@@ -1,10 +1,25 @@
     private final Name name;
     private int currentPosition = 0;
 
+    @Override
+    public boolean equals(Object o) {
```
`Object`클래스의 오버라이드 함수인 `equals` `hashCode` `toString`은 일반적으로
클래스 맨 마지막에 구현합니다. 참고하세요😄

gw-study commented on src/main/java/racingcar/Winners.java:
```diff
@@ -2,30 +2,23 @@
 
-    public List<String> findWinnerCarNames(){
-        return winnerNames();
+    public List<String> findWinnerCarNames(List<Car> winnersCar){
```
`findWinnerCarNames`와 `winnerNames`함수의 차이는 무엇일까요?🤔
중복은 아닐까요?

gw-study commented on src/main/java/racingcar/Winners.java:
```diff
@@ -38,4 +31,10 @@ public int getMaxPosition(List<Car> readyCars){
     }
+
+    public List<Car> winnerCars() {
+        return cars.stream()
+                .filter(car -> car.getPosition() == getMaxPosition(cars))
+                .collect(Collectors.toList());
```
스트림 사용 잘해주셨어요👍

gw-study commented on src/test/java/racingcar/RacingCarTest.java:
```diff
@@ -54,9 +54,9 @@ public void getWinnersName(){
+        Winners winners = new Winners(cars);
 
-        assertThat(String.join(",", racing.getWinners().findWinnerCarNames())).isEqualTo("pobi");
+        assertThat(winners.winnerCars()).containsExactly(car1);
```
테스트도 적절하게 잘 수행주셨습니다👏

gw-study reviewed (approved):
피드백 적용 잘해주셨습니다.👏👏👏

몇가지 소소한 코멘트 드린게 있는데 과정 진행하시면서 신경써 주시면 될거 같아요😄

이번 단계는 여기서 머지하도록 할께요

여기까지 진행하느라 많이 힘드셨을텐데 드디어 마지막 단계이네요 :)
마지막 단계는 리팩터링 단계인데요,
요구사항(패키지 분리) 신경 쓰시면서 다시 한번 코드를 점검해 보고
리팩터링이 필요한 부분이 있을까 고민하시면서 진행해봐 주시면 좋을거 같습니다

그럼 다음 단계에서 만나요😁🚀
